### DISCUSSION TITLE
Periodic and/or UI initiated reindex

### DISCUSSION THREAD

**Discussion by officerNordberg** [DependencyTrack/dependency-track#1641] [Ideas]
I noticed my vulnerable software index stopped updating at various times in the last month on all 3 of my instances (prod, test, local-dev).  The only way I know to fix this is stop DT, delete the index files and restart it.  Is anyone else running into this?  Can this be made more resilient?  I was going to start looking into ways to detect that the index was stale or possibly investigate ways make vulnerability updates more transactional between the DB and lucene index.  It's possible I'm only running into this because of my work on fuzzy matching but I've also noticed lucene indexes can end up incomplete or corrupt if the system is interrupted during it's initial startup.  Any thoughts?

**Comment by nscuro** [comment]
> I noticed my vulnerable software index stopped updating at various times in the last month on all 3 of my instances (prod, test, local-dev).

Were the indexes not updating (as in, the actual files not changing) or was it that the results returned by `IndexSearcher` were stale? If it's the latter, then it may be related to https://github.com/DependencyTrack/dependency-track/pull/1612 and should be fixed in 4.5.0.

**Reply by officerNordberg** [reply]
The actual files were not updating. They were last updated April 8th in one instance, I don't recall the last dates on the other 2 systems which I already rebuilt while triaging the issue. This was noticed because we have another third-party system that reports on vulnerabilities based on our SBOM that we use to validate our DT results that started showing vulnerabilities that were not also reflected in all relevant DT projects.   

For example: 

CVE-2022-25647 was published May 1st.  It was in our local database.  It is not in the OSS Index so the PURL is useless. 

It was located by DT projects whose SBOMs were created by Syft because CPE is supplied and the CPE Part Vendor is used instead of the erroneous com.google.code.gson
https://github.com/DependencyTrack/dependency-track/blob/a3b42a6fe4d19fc91122887dba0841f26fa7c7c4/src/main/java/org/dependencytrack/tasks/scanners/InternalAnalysisTask.java#L128
`
   {
      "group" : "com.google.code.gson",
      "name" : "gson",
      "version" : "2.7",
      "cpe" : "cpe:2.3:a:google:gson:2.7:*:*:*:*:*:*:*",
      "purl" : "pkg:maven/com.google.code.gson/gson@2.7",
      "type" : "library",
      "bom-ref" : "f2378f93-5352-45c3-a234-4d4ec2e445ce"
    }
` 
Projects whose SBOMs were created by Gradle don't match at all without a CPE and also don't match using my WIP fuzzy matching if the VulnerableSoftware Lucene index is out of date 

`
    {
      "group" : "com.google.code.gson",
      "name" : "gson",
      "version" : "2.2.2",
      "description" : "Google Gson library",
      "hashes" : [],
      "purl" : "pkg:maven/com.google.code.gson/gson@2.2.2?type=jar",
      "externalReferences" : [],
      "type" : "library",
      "bom-ref" : "d44f014c-5ab5-4eca-b22e-f82ce45a5837"
    },
`
Even without the CPE the fuzzy matching would search for cpe23 matching `cpe:2.3:a:*:gson:*:*:*:*:*:*:*:*` but like I said it was not in the index, just the database.

So there's a lot going on here.  The Gradle CycloneDX plugin could be improved to include a very helpful CPE and in general just spit balling on how DT could be made to recognize when the database and lucene index are out-of-sync for whatever reason.  I'm still investigating that.  I'm going to make sure it's not the fuzzy matching searches themselves that are locking the index somehow but outside of that I'm not sure where to start the logs don't have any smoking gun so far.

**Reply by officerNordberg** [reply]
I should mention, besides looking at the file dates, I used Luke to validate the actual contents of the Lucene Index while triaging

**Comment by officerNordberg** [comment]
I'm still having issues with the index getting stale and needing to be recreated.  I've speculated a lot of the open issues I see come across about DT not finding Vulnerabilities might be connected to this as you'll never be able to recreate it locally.  For the fourth time this year I recently stopped the server, sudo rm -rf dependency-track/index/vulnerablesoftware/, restarted and suddenly vulnerabilities are found.  These were observable in the database prior to shutdown and removal of the lucene index.   I've been instructed to write a job to do this on a schedule from now on but I really think this should be addressed for DT as a whole.  I'm not sure how many other people running DT have another source to independently validate their results like we do. but it's pretty clear to me that for whatever reason the index degrades over time and results in fewer findings.
